### DISCUSSION TITLE
minimal-racket encounters SSL handshake failure when using raco

### DISCUSSION THREAD

**Discussion by rami3l** [Homebrew/discussions#180] [Everyday usage]
`minimal-racket` is supposed to be a minimal distribution with more components to be added with `raco pkg install` command. However, after running `brew install minimal-racket`, I found that `raco pkg` is nearly unusable, because it cannot establish a connection to the main package repository:

<details><summary>Error Log</summary>
<p>

```bash
raco pkg catalog-show racket-langserver
port->ssl-ports: connect failed (error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure)
  context...:
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/openssl/mzssl.rkt:1466:0: ports->ssl-ports
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/net/http-client.rkt:280:0: http-conn-CONNECT-tunnel
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/racket/contract/private/arrow-val-first.rkt:555:3
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/net/url.rkt:166:0: make-ports
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/net/url.rkt:201:0: http://getpost-impure-port
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/net/url.rkt:309:2: redirection-loop
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/racket/contract/private/arrow-val-first.rkt:555:3
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/pkg/private/network.rkt:59:3
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/pkg/private/catalog.rkt:218:0: read-from-server
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/pkg/private/catalog.rkt:138:9: for-loop
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/pkg/private/catalog.rkt:135:2: lookup-normally
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/pkg/private/catalog.rkt:294:0: get-pkg-details-from-catalogs
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/pkg/private/catalog-show.rkt:36:4: for-loop
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/pkg/private/catalog-show.rkt:11:0: pkg-catalog-show
   /usr/local/Cellar/minimal-racket/7.9/share/racket/collects/racket/contract/private/arrow-val-first.rkt:555:3
   (submod "/usr/local/Cellar/minimal-racket/7.9/share/racket/collects/pkg/main.rkt" main): [running body]
   ...
```

</p>
</details>

PS: I cannot be sure if this problem is reproducible or not, so I started the discussion here instead of posting an issue.

**Comment by SMillerDev** [comment]
What made you suspect that it's a packaging issue?

**Reply by rami3l** [reply]
I'm not sure, but I've noticed that:

- This issue should have nothing to do with network situation (I have another Linux box under the same network and it installs alright).
- The official precompiled image (minimal variant) seems to work.
- I have considered the probability of misconfigured package catalog, but unfortunately a reset to default (`raco pkg config --set catalogs https://download.racket-lang.org/releases/7.9/catalog/ ""`) didn't solve this problem.

**Reply by rami3l** [reply]
Sorry for such a long delay, but now I'm sure this has something to do with `libssl`.
As stated [in the Racket docs](https://docs.racket-lang.org/openssl/), on macOS the [minimal Racket distribution](https://download.racket-lang.org/releases/8.0/installers/racket-minimal-8.0-x86_64-macosx-cs.tgz) is shipped with `racket/lib/libssl.1.1.dylib` (`raco` in this tarball works quite well), which is somehow missing in your package 🤔
I'm rather new to `brew`, and just being curious, what makes the `brew` distribution so different from the official precompiled version?

**Reply by gromgit** [reply]
As the docs you pointed to say:
> For Mac OS, openssl depends on "libssl.dylib" and "libcrypto.dylib". Although those libraries are provided by Mac OS 10.2 and later, their use is deprecated, so the Racket distribution for Mac OS includes newer versions.

Homebrew has an up-to-date OpenSSL bottle, so minimal-racket should be built to use that, instead of vendoring its own OpenSSL libraries from a third-party source.

That said, minimal-racket 8.0 was released since your initial post, and I was able to run `raco pkg catalog-show racket-langserver` without errors on Big Sur. Are you still seeing the same issue? If so, post the output of `brew config` and `brew doctor`.

**Reply by carlocab** [reply]
It also looks like `minimal-racket` is using the source + built packages archive, which contains pre-built artifacts. Sounds like we shouldn't be doing that.

**Reply by gromgit** [reply]
Hmmm, let me see if it'll build from pure source.

**Reply by gromgit** [reply]
`brew install` and `brew test` run OK from pure source, but the generated bytecode files (`*.zo`) are all different sizes from the pre-built ones in the current bottle (which are of course identical to those in the source+builtpkgs tarball).

Also, when I select "Unix Source" from https://download.racket-lang.org/, the following recommendation is presented:
> **About source distributions:**
> The **Source + built packages** distribution is recommended, instead of the selected **Source** distribution. The **Source + built packages** distribution includes pre-built, platform-independent bytecode; it installs much faster than plain source, and it is also compatible with fast installs of additional Racket packages.

Since I have zero experience with Racket, I'm reluctant to push this "pure source" build through as a PR.

**Reply by carlocab** [reply]
How large is the diff on the formula? I'll open the PR if you'd rather not.

**Reply by gromgit** [reply]
It's just the usual `url`/`sha256`/`revision` bump. I guess I might as well push it through, then see if any Racket users start complaining of sudden failures. :grin:

**Reply by rami3l** [reply]
@gromgit After `minimal-racket/8.0_2` this problem seems to have gone... Thanks a lot!

**Comment by rami3l** [comment]
For the record: the reason is still unclear, but https://github.com/NixOS/nixpkgs/issues/65337 might be relevant.
